lib/src/bitarray.dart

```diff
@@ -0,0 +1,317 @@
+  static BitArray andGate(BitArray ba1, BitArray ba2) {
+    _checkLengthsEqual(ba1, ba2);
+    var result = BitArray(ba1.length);
+    result._setDataWords((int wordIndex) {
+      return ba1._data[wordIndex] & ba2._data[wordIndex];
+    });
+    return result;
+  }
+
+  /// Returns a result of logical XOR operation for passed [BitArray]s.
+  static BitArray xorGate(BitArray ba1, BitArray ba2) {
+    _checkLengthsEqual(ba1, ba2);
+    var result = BitArray(ba1.length);
+    result._setDataWords((int wordIndex) {
+      return ba1._data[wordIndex] ^ ba2._data[wordIndex];
+    });
+    return result;
+  }
+
+  /// Returns [true] if bitarrays have the same length and the same values.
+  bool equal(BitArray ba) {
+    if (identical(this, ba)) return true;
+    if (_length != ba._length) return false;
+    for (var wordIndex = 0; wordIndex < _data.length; ++wordIndex) {
+      if (_data[wordIndex] != ba._data[wordIndex]) return false;
+    }
+    return true;
+  }
+
+  @override
+  bool operator [](int index) {
+    RangeError.checkValidIndex(index, this);
+    return (_data[index >> 5] & (1 << (index & 31))) != 0;
+  }
+
+  @override
+  int get length => _length;
+
+  @override
+  void operator []=(int index, bool val) {
+    RangeError.checkValidIndex(index, this);
+    if (val) {
+      _data[index >> 5] |= 1 << (index & 31);
+    } else {
+      _data[index >> 5] &= ~(1 << (index & 31));
+    }
+  }
+
+  /// Counts number of bits in unsigned [val].
+  int _countBits(int val) {
+    var bits = 0;
+    while (val != 0) {
+      bits += _numOfBitsLookup[val & 0xFF];
+      val >>= 8;
+    }
+    return bits;
+  }
+
+  /// Returns number [true] values in [this].
+  int get cardinality {
+    var card = 0;
+    for (var word in _data) {
+      card += _countBits(word);
+    }
+    return card;
+  }
+
+  /// Returns logical NOT of [this].
+  BitArray operator ~() => BitArray.notGate(this);
+
+  /// Returns result or logical AND between [this] and [ba].
+  BitArray operator &(BitArray ba) => BitArray.andGate(this, ba);
+
+  /// Returns result or logical OR between [this] and [ba].
+  BitArray operator |(BitArray ba) => BitArray.orGate(this, ba);
+
+  /// Returns result or logical XOR between [this] and [ba].
+  BitArray operator ^(BitArray ba) => BitArray.xorGate(this, ba);
+
+  /// Checks lengths equal for given [BitArray]s.
+  ///
+  /// Throws [ArgumentExceprion] when lengts is differ.
+  static void _checkLengthsEqual(BitArray ba1, BitArray ba2) {
+    if (ba1._length != ba2._length) {
+      throw ArgumentError.value(
+          'BitArrays should have equal lengths: ${ba1._length} != ${ba2._length}');
+    }
+  }
+
+  /// Sets the bit at the [index] to the complement of its current value.
+  void flip(int index) {
+    RangeError.checkValidIndex(index, this);
+    _data[index >> 5] ^= 1 << (index & 31);
+  }
+
+  /// Sets each bit at the range from [start] inclusive to [end] exclusive to the complement of its current value.
```

Comment:
Shorter first sentence, then elaborate below. Maybe:

/// Toggles each bit in the range from [start] to [end].
///
/// Sets each bit in the range from [start] (inclusive) to [end] (exclusive) to the complement
/// of its current value.
///
/// The [start] and [end] must define a valid range of this list (0 &le; [start] &le; [end] &le; [length]).
/// If [end] is omitted, it defaults to [length].